178 datacollection from userdict wbaccinelli

psomhorst opened the pull request:
This bug fixes (?) the equality and equivalence checks for DataCollections introduced in #182.

Because the formal tests have not been entirely rewritten this is not extensively tested yet. Equivalence and equality were both tested ad hoc, as well as concatenation of two DataCollections.

Also adds concatenation of Sequence.continuous_data and Sequence.sparse_data (required for testing this). 

I propose merging this now to prevent bugs blocking usage of main.

wbaccinelli reviewed (approved):
Looks good to me

DaniBodor commented on eitprocessing/datahandling/mixins/equality.py:
```diff
@@ -13,9 +14,10 @@ class Equivalence:
     # inspired by: https://stackoverflow.com/a/51743960/5170442
-    def __eq__(self, other: Self):
+    def __eq__(self, other: Self) -> bool:
         if self is other:
             return True
+
         if is_dataclass(self):
             if self.__class__ is not other.__class__:
                 return NotImplemented
```
I think raising `NotImplemented` should happen whenever classes of two objects don't match, not only if we're dealing with data classes.
Also, what is the reason to check the `__class__` instead of the `type()` (as done in line 81)?

```suggestion
        if self is other:
            return True
        if self.__class__ is not other.__class__:
            return NotImplemented

        if is_dataclass(self):
```

psomhorst replied:
I think `NotImplemented` is not the right result here anyway. If the types don't match, dataclass or not, `__eq__` should return False. Or am I missing something?

I changed it to return `False` in a968c9daa4bf22c36b616a310d9b0c8e69b8436d.

DaniBodor commented on eitprocessing/datahandling/mixins/equality.py:
```diff
@@ -24,6 +26,10 @@ def __eq__(self, other: Self):
                 return False
             return all(Equivalence._array_safe_eq(a1, a2) for a1, a2 in zip(t1, t2, strict=True))
+
+        if isinstance(self, UserDict) and isinstance(other, UserDict):
+            return all(Equivalence.__eq__(self[key], other[key]) for key in self)
```
Above comment alleviates the need to check that both are `UserDict` (because if only one is, the method would return `None`)

```suggestion
        if isinstance(self, UserDict):
            return all(Equivalence.__eq__(self[key], other[key]) for key in self)
```

psomhorst replied:
That is correct.

There was a flaw in the logic here. It assumed that other has all the keys self had, but didn't check. Also, it didn't check whether other had any keys self didn't have.

I updated in in 294e06bf839f9a328d107b45ca72341adb7ddf73 to check all keys in self and other. If one of them doesn't have the key, the method returns `False`.

DaniBodor reviewed (commented):
Thanks for picking this up!
I have a few comments/questions

psomhorst commented on eitprocessing/datahandling/mixins/equality.py:
```diff
@@ -77,7 +83,7 @@ def isequivalent(self, other: Self, raise_: bool = False) -> bool:  # noqa: C901
 
             # check keys in collection
-            if isinstance(self, dict):
+            if isinstance(self, dict | UserDict):
```
That is a valid question. I don't think so, but am currently unable to test it.

I'll add a TODO and issue to work on after the tests are finished.

psomhorst commented:
@DaniBodor Thanks for the great comments. 

I addressed all the comments. I added some logic as well, seeing that there were some things missing in __eq__.

DaniBodor commented on eitprocessing/datahandling/mixins/equality.py:
```diff
@@ -28,8 +28,12 @@ def __eq__(self, other: Self) -> bool:
 
-        if isinstance(self, UserDict) and isinstance(other, UserDict):
-            return all(Equivalence.__eq__(self[key], other[key]) for key in self)
+        if isinstance(self, UserDict):
+            try:
+                keys_from_both = set(self.keys()) | set(other.keys())
+                return all(Equivalence.__eq__(self[key], other[key]) for key in keys_from_both)
+            except KeyError:
+                return False
```
isn't this doing the same but simpler?

```suggestion
        if isinstance(self, UserDict):
            if set(self.keys()) != set(other.keys()):
                return False
            return all(Equivalence.__eq__(self[key], other[key]) for key in set(self.keys()))
```

DaniBodor commented:
I rebased this onto main, so that I can rebase #164 onto this without causing weird double conflicts
